crates/vm/src/system/memory/paged_vec.rs

```diff
@@ -44,36 +44,109 @@ impl<T: Default + Clone, const PAGE_SIZE: usize> PagedVec<T, PAGE_SIZE> {
             None
         }
     }
-
     #[inline(always)]
     pub fn range_vec(&self, range: Range<usize>) -> Vec<T> {
-        let mut result = Vec::with_capacity(range.len());
-        for page_idx in (range.start / PAGE_SIZE)..range.end.div_ceil(PAGE_SIZE) {
-            let in_page_start = range.start.saturating_sub(page_idx * PAGE_SIZE);
-            let in_page_end = (range.end - page_idx * PAGE_SIZE).min(PAGE_SIZE);
-            if let Some(page) = self.pages[page_idx].as_ref() {
-                result.extend(page[in_page_start..in_page_end].iter().cloned());
+        let len = range.end - range.start;
+        // Create a vector for uninitialized values.
+        let mut result: Vec<MaybeUninit<T>> = Vec::with_capacity(len);
+        // SAFETY: We set the length and then initialize every element.
+        unsafe {
```

Comment:
Can we put the `unsafe` blocks *only* around the parts that use unsafe APIs? This will make it much easier to reason about safety.

Reply:
it does seem like nearly the whole block is unsafe

Reply:
maybe it can be split up into a few more blocks though